src/scanner.js

```diff
@@ -0,0 +1,69 @@
+import taskcluster from 'taskcluster-client'
+import scopeUtils from 'taskcluster-lib-scopes'
+import User from './user'
+var debug = require('debug')('scanner');
+
+export default async function scanner(cfg, authorizers) {
+  // * get the set of identityProviderIds
+  // * for each:
+  //   * fetch all clients, sort by identity
+  //   * for each identity:
+  //     * get roles from providers, expand
+  //     * for each client in that identity:
+  //       * get, verify client.expandedScopes satisfied by identity's expandedScopes
+
+  // NOTE: this function performs once auth operation at a time.  It is better
+  // for scans to take longer than for the auth service to be overloaded.
+  let auth = new taskcluster.Auth({credentials: cfg.app.credentials});
+
+  // gather all identityProviderIds used in any authorizer
+  let identityProviders = new Set(
+      authorizers.map(authz => authz.identityProviders).reduce((a,b) => a.concat(b)))
+
+  // enumerate all clients for any of those identity providers
+  let clients = [];
+  for (let idp of identityProviders) {
+    clients = clients.concat(await auth.listClients({prefix: idp + "/"}));
+  }
+  clients.sort((a, b) => {
+    if (a.clientId < b.clientId) return -1;
+    if (a.clientId > b.clientId) return 1;
+    return 0;
+  });
+
+  // iterate through the clients, constructing a new User as necessary, comparing
+  // the client's scopes to the User's scopes and disabling where necessary.
+  let user, userScopes;
+  let idPattern = /^([^\/]*\/[^\/]*)\/.+$/
+  for (let client of clients.sort()) {
+    debug("examining client", client.clientId);
+    if (!client.clientId.match(idPattern) || client.disabled) {
+      continue;
+    }
+
+    // refresh the user if it does not correspond to this client
+    let clientIdentity = client.clientId.replace(idPattern, '$1');
+    if (!user || user.identity != clientIdentity) {
+      user = new User();
+      user.identity = clientIdentity;
+
+      await Promise.all(authorizers.map(authz => {
+        if (authz.identityProviders.indexOf(user.identityProviderId) !== -1) {
+          return authz.authorize(user)
+        }
+      }));
+
+      userScopes = (await auth.expandScopes({scopes: user.scopes()})).scopes;
+      // allow the implicit 'assume:client-id:<clientId> auth adds for each client
+      userScopes.push('assume:client-id:' + clientIdentity + '/*');
```

Comment:
No, that wouldn't make sense, since `expandScopes` doesn't know what it's expanding scopes for.  To be honest, I'd like to get rid of the `assume:client-id` bit entirely -- it's not currently used.